### ISSUE TITLE
Unexpected/incorrect string encoding in some responses from Kodi

### ISSUE THREAD

**Issue by jingai** [m0ngr31/kodi-alexa#172]
This is best explained by example.

I have an album called `La Maison De Mon Rêve`.

If I manually select a song from this album to play and then use Player.GetItems to see what's playing, I get `u'La Maison De Mon R\xeave'`.

If I allow the player to advance to the next song on the album on its own, Player.GetItems gives me `u'La Maison De Mon R\u0102\u015eve'`.

If I encode the latter string in UTF-8, I get `La Maison De Mon RĂŞve`.  But if I encode it in ISO-8859-2, I get `La Maison De Mon Rêve`.

chardet.detect() backs up the above as well.

From what I've read on the Kodi forums and docs, it's supposed to _always_ encode strings in UTF-8.  For some reason it's not for us (sometimes).

@mcl22 noted a similar problem with the genre (in German) `Komödie`.  I'm guessing the scraper decided to store this in some encoding other than UTF-8 too.  The web-based slot generator likely doesn't show this issue simply because the user's web browser can detect the encoding and display it correctly.  We _do_ however see a problem with `Komödie` in the skill later -- even if the string is encoded correctly in the slots -- because we encode it as UTF-8 in requests that filter on genre, but Kodi is matching against the ISO-8859-2 string it has in the database.

I _think_ this is fundamentally a problem in Kodi (I'm using 17.3).  We _might_ want to work around it by detecting the encoding, though we'd in theory have to do that in a lot of places to cover all the bases.

Because the docs and Kodi forum posts suggest all strings will be UTF-8, I'm guessing this kind of detection and conversion is supposed to happen there.  But it's not in all cases.

Therefore, I don't think this is an actionable item for us, but since it took me a bit of playing around to figure out what was going on, I figured I'd at least document it here.

**Comment by mcl22** [comment]
Not sure if this is the right topic but it could be :) While playing around with the utterances I've just tried the WhatNewMovies. That works almost perfectly (in german). But when I try to ask for new "Komödie" (comedy) I get this error:
Trying to match: komodie
Simple match failed, trying fuzzy match...
Trying to match komodie from <function digits2roman at 0x0355BC70>
Fuzzy match 92%
Neu hinzugef\xc3\xbcgte kom\xc3\xb6die Filme
Sending request to http://127.0.0.1:8081/jsonrpc from device amzn1.ask.device.AH***
[2017-05-27 17:14:50,714] ERROR in app: Exception on / [POST]
 Traceback (most recent call last):
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1982, in wsgi_app
     response = self.full_dispatch_request()
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1614, in full_dispatch_request
     rv = self.handle_user_exception(e)
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1517, in handle_user_exception
     reraise(exc_type, exc_value, tb)
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1612, in full_dispatch_request
     rv = self.dispatch_request()
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1598, in dispatch_request
     return self.view_functions[rule.endpoint](**req.view_args)
   File "C:\\Python27\\lib\\site-packages\\flask_ask\\core.py", line 594, in _flask_view_func
     result = self._map_intent_to_view_func(self.request.intent)()
   File "C:/***/htdocs/kodi-alexa\\alexa.py", line 2054, in alexa_what_new_movies
     new_movies = kodi.GetUnwatchedMoviesByGenre(genre_located['label'])
   File "C:\\Python27\\lib\\site-packages\\kodi_voice\\kodi.py", line 1119, in GetUnwatchedMoviesByGenre
     for d in data['result']['movies']:
 KeyError: 'movies'

**Comment by jingai** [comment]
No, that is not related to this Issue -- your problem there is that Kodi didn't return any movies that match that criteria.

**Comment by mcl22** [comment]
Not sure if it's the right topic but I encountered an error while testing the ListenToAlbum intend (onyl asking for album - without artist). It performs great but for 2 albums I get the following error:
Spiele Album
 ERROR in app: Exception on / [POST]
 Traceback (most recent call last):
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1982, in wsgi_app
     response = self.full_dispatch_request()
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1614, in full_dispatch_request
     rv = self.handle_user_exception(e)
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1517, in handle_user_exception
     reraise(exc_type, exc_value, tb)
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1612, in full_dispatch_request
     rv = self.dispatch_request()
   File "C:\\Python27\\lib\\site-packages\\flask\\app.py", line 1598, in dispatch_request
(**req.view_args)
   File "C:\\Python27\\lib\\site-packages\\flask_ask\\core.py", line 594, in _flask_view_func
     result = self._map_intent_to_view_func(self.request.intent)()
   File "C:/***/htdocs/kodi-alexa\\alexa.py", line 642, in alexa_listen_album
     album_id, album_label = kodi.FindAlbum(Album)
   File "C:\\Python27\\lib\\site-packages\\kodi_voice\\kodi.py", line 454, in FindAlbum
     print 'Searching for album "%s"' % (sanitize_name(heard_search))
   File "C:\\Python27\\lib\\site-packages\\kodi_voice\\kodi.py", line 26, in sanitize_name
     name = unicodedata.normalize('NFKD', media_name).encode('ASCII', 'ignore')
 TypeError: normalize() argument 2 must be unicode, not None

This is for the albums "weiß" and "schwarz" by Böhse Onkelz. The app history tells:
alexa öffne test und spiele das album weiß
alexa öffne test und spiele das album schwarz
In the simulator both work. But then the log states:
Searching for album "Schwartz"
Searching for album "wei"

If I use ask for album by artist both, "schwarz" and "weiß" work.

**Comment by jingai** [comment]
I think I just accidentally solved this.

@mcl22 reported that responses from Kodi were taking an abnormally long time.  After a bit of digging, I found that using requests.json() instead of json.loads(r.text) solved the problem.  The issue was that if requests couldn't assume the response was JSON, it couldn't assume the character encoding either, and thus had to detect it.

...which got me to thinking that _this_ issue is probably when it detects it as something other than UTF-8.  And sure enough, the problem appears to be gone for the cases I've experienced.

I'm going to verify with @mcl22 before closing this.  He noted a title that got mangled and as well, problems requesting `Komödie` movies.
